package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,333 @@
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    exit(EXIT_FAILURE);
+  }
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_setopt(curl, CURLOPT_URL, config->endpoint_url);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &download_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &config->fd_out);
+#ifdef VERBOSE
+  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+#endif
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+  struct curl_slist *chunk = NULL;
+  chunk = curl_slist_append(chunk, config->encoding);
+  chunk = curl_slist_append(chunk, config->accept_type_header);
+  chunk = curl_slist_append(chunk, config->device_header);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+  res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
+    log_error(curl_easy_strerror(res));
+    curl_easy_cleanup(curl);
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_cleanup(curl);
+  curl_global_cleanup();
+  return NO_ERROR;
+}
+
+static void base_obs_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_OBS, len, msg);
+}
+
+static void base_llh_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_LLH, len, msg);
+}
+
+static void base_ecef_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_ECEF, len, msg);
+}
+
+u32 msg_read(u8 *buf, u32 n, void *context)
+{
+  int *fd = (int *)context;
+  ssize_t m = read(*fd, buf, n);
+  return m;
+}
+
+RC download_io_loop(skylark_client_config_t *config)
+{
+  // TODO sender id is not going to be right here, since we're relaying
+  // messages.
+  sbp_zmq_state_t sbp_zmq_state;
+  if (sbp_zmq_init(&sbp_zmq_state, config->sbp_zmq_config) < 0) {
+    exit(EXIT_FAILURE);
+  }
+  sbp_state_t sbp_state;
+  sbp_msg_callbacks_node_t callback_node;
+  sbp_state_init(&sbp_state);
+  sbp_state_set_io_context(&sbp_state, &config->fd_in);
+  sbp_register_callback(&sbp_state, SBP_MSG_OBS,
+                        &base_obs_callback, &sbp_zmq_state,
+                        &callback_node);
+  sbp_msg_callbacks_node_t base_ecef_callback_node;
+  sbp_register_callback(&sbp_state, SBP_MSG_BASE_POS_ECEF,
+                        &base_ecef_callback, &sbp_zmq_state,
+                        &base_ecef_callback_node);
+  sbp_msg_callbacks_node_t base_llh_callback_node;
+  sbp_register_callback(&sbp_state, SBP_MSG_BASE_POS_LLH,
+                        &base_llh_callback, &sbp_zmq_state,
+                        &base_llh_callback_node);
+  for (;;) {
+    sbp_process(&sbp_state, &msg_read);
```

Comment:
sbp process returns an error code - I was processing it, but it broker when I exited on a failure.